service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java

```diff
@@ -219,6 +222,43 @@ public Integer getNumberOfSamplesInMolecularProfile(String molecularProfileId) {
         return molecularDataList;
     }
 
+    @Override
+    public List<GeneMolecularData> getMolecularDataInMultipleMolecularProfilesByGeneQueries(List<String> molecularProfileIds,
+                                                                                            List<String> sampleIds,
+                                                                                            List<GeneFilterQuery> geneQueries,
+                                                                                            String projection) {
+        // Molecular alterations for all genes in the geneQueries
+        List<Integer> entrezGeneIds = geneQueries.stream().map(q -> q.getEntrezGeneId()).collect(Collectors.toList());
+        List<GeneMolecularData> molecularDataList = getMolecularDataInMultipleMolecularProfiles(molecularProfileIds, sampleIds, entrezGeneIds, projection);
+
+        // All CNA events that match requested geneQueries
+        List<DiscreteCopyNumberData> copyNumberData = discreteCopyNumberRepository
+            .getDiscreteCopyNumbersInMultipleMolecularProfilesByGeneQueries(molecularProfileIds, sampleIds,
+                geneQueries, projection);
+        
+        // molecularProfile->entrezGeneId->sampleId->alterationType lookup table for CNA events
+        Map<String, Map<Integer, Map<String, Map<Integer, String>>>> cnaEventLookup = copyNumberData.stream()
```

Comment:
instead of multi-grouping why not create multikey map?

Reply:
Yes, much nicer. Thanks for this suggestion. So updated...